crates/typed-store/src/sally/mod.rs

```diff
@@ -0,0 +1,588 @@
+    where
+        J: Borrow<K>,
+        U: Borrow<V>,
+    {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => db_map.multi_insert(key_val_pairs),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => test_db.multi_insert(key_val_pairs),
+        }
+    }
+    fn multi_remove<J>(&self, keys: impl IntoIterator<Item = J>) -> Result<(), Self::Error>
+    where
+        J: Borrow<K>,
+    {
+        self.batch().delete_batch(self, keys)?.write()
+    }
+    fn try_catch_up_with_primary(&self) -> Result<(), Self::Error> {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => Ok(db_map.try_catch_up_with_primary()?),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => Ok(test_db.try_catch_up_with_primary()?),
+        }
+    }
+}
+
+impl<J, K, U, V> TryExtend<(J, U)> for SallyColumn<K, V>
+where
+    J: Borrow<K> + std::clone::Clone,
+    U: Borrow<V> + std::clone::Clone,
+    K: Serialize,
+    V: Serialize,
+{
+    type Error = TypedStoreError;
+
+    fn try_extend<T>(&mut self, iter: &mut T) -> Result<(), Self::Error>
+    where
+        T: Iterator<Item = (J, U)>,
+    {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => db_map.try_extend(iter),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => test_db.try_extend(iter),
+        }
+    }
+    fn try_extend_from_slice(&mut self, slice: &[(J, U)]) -> Result<(), Self::Error> {
+        match self {
+            SallyColumn::RocksDB((
+                db_map,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => db_map.try_extend_from_slice(slice),
+            SallyColumn::TestDB((
+                test_db,
+                SallyConfig {
+                    mode: SallyRunMode::FallbackToDB,
+                },
+            )) => test_db.try_extend_from_slice(slice),
+        }
+    }
+}
+
+/// A Sally write batch provides a mutable struct which holds a collection of db mutation operations and
+/// applies them atomically to the db.
+pub enum SallyWriteBatch {
+    // Write batch for RocksDB backend when `fallback_to_db` is set as true
+    RocksDB(DBBatch),
```

Comment:
Can you say something about how a multi-db batch might work?

I also wonder if batching should really be exposed to the client code so directly. The client needs to have a unit of atomicity, but how that is achieved might best be abstracted away. For instance, we might want to combine multiple independent "write groups" into a single batch (or one single batch per db) for efficiency.

Reply:
I removed the insert/remove/ etc methods from the column and now the only way to update a db is through write batch